config-sample.yaml

```
@@ -112,6 +112,15 @@ rtc:
   # batch_io:
   #    batch_size: 128
   #    max_flush_interval: 2ms
+  # # force a reconnect on a publication error
+  # reconnect_on_publication_error: true
+  # # force a reconnect on a subscription error
+  # reconnect_on_subscription_error: true
+  # # force a reconnect on a data channel error
+  # reconnect_on_data_channel_error: true
+  # # max number of bytes to buffer for data channel. 0 means unlimited.
+  # # when this limit is reached, will drop data messages and also issue a data channel error for reliable channel
```

```suggestion
  # # when this limit is breached, data messages will be dropped till the buffered amount drops below this limit. this applies to both reliable and unreliable data channel. In addition, for reliable data channel only, forces a reconnect if `reconnect_on_data_channel_error: true`
```
Don't necessarily like the language I suggested as I think it can be made clearer. Main thing I wanted to convey
1. Limit applies to both unreliable and reliable data channels.
2. Don't need to mention that exceeding the limit throws an internal error.
3. Data messages will flow again when the buffered amount falls below the limit.
4. A full reconnect can be configured for this condition (hence the suggestion to point at the other config `reconnect_on_data_channel_error`) and that applies to reliable data channel only.

---

Who should we force to disconnect? the publisher or the subscriber?

---

I think subscriber as that is the slow consumer.

---

Ok, from the PR I can tell it's the subscriber.. to be honest, this behavior is a bit confusing.

This means a malicious user in the room could force everyone to disconnect by spamming the room with reliable data. I'm not sure if the alternative is better (i.e. disconnect the sender).. it might be better to not force disconnect, but to break the contract of reliable delivery.. (since it's not guaranteed delivery today)

---

It will break reliable delivery only unless the other config `reconnect_on_data_channel_error` is enabled. We can remove that forcing a reconnect and leave it at dropping messages when limit breached.

---

Disconnect the sender is hard, it will confuse the publisher if a slow consumer exists in a large room and limit the rate of publish message by the slowest subscriber.

---

I don't think disconnecting sender is the right thing either.. thinking about it.. I don't think disconnecting either client is the correct course of action (yes, even with a config parameter.)

---

There seems to be no ideal way of handling this. We have to choose between:
* disconnect publisher
* disconnect slow receiver
* do not disconnect either (break reliable guarantees)

The reason why I'm now leaning against 3 is because it seems like the least "bad" option. and basically we are giving it to users by default anyways (because `reconnect_on_data_channel_error` is not enabled by default)

---

Let's just keep it simple for this iteration then. Just drop messages. No reconnect. The config for max buffered amount is applied to drop/let messages through. No other change. We can think about disconnect more.